zerver/lib/users.py

```diff
@@ -342,8 +342,59 @@ def access_user_by_email(
+    1) The real email address of the intended user, if the requester believes that user exists and allows
+       their email address to be visible to the requester.
+    2) The dummy email address (of the approximate shape 'user<id>@<dummy email domain>') of the intended
+       user. We detect when the format of the provided email address matches the format of our dummy
+       email addresses and extract the user id from it for a regular id-based lookup.
+
+       In particular, this mode is kept around for backwards compatibility with the old behavior of the
+       `GET /users/{email}` endpoint, which required use of the dummy email address for lookups of any user
+       which didn't have EMAIL_ADDRESS_VISIBILITY_EVERYONE set, regardless of how the actual
+       email_address_visibility setting related to the role of the requester.
+
+       Note: If the realm.host value changes (e.g. due to the server moving to a new
+       domain), the required dummy email values passed here will need to be updated
+       accordingly to match the new value. This deviates from the original API behavior,
+       where the lookups were supposed to match the UserProfile.email value, which was
+       **not** updated for existing users even if the server moved domains. See
+       get_fake_email_domain for details of how the email domain for dummy email addresses
+       is determined.
+
+    The purpose of this is to be used at API endpoints that allow selecting the target user by
+    delivery_email, while preventing the endpoint from leaking information about user emails.
+    """
+
+    # First, check if the email is just the dummy email address format. In that case,
+    # we don't need to deal with email lookups or email address visibility restrictions
+    # and we simply get the user by id, extracted from the dummy address.
+    dummy_email_regex = get_dummy_email_address_for_display_regex(user_profile.realm)
+    match = re.match(dummy_email_regex, email)
+    if match:
+        target_id = int(match.group(1))
+        return access_user_by_id(
+            user_profile,
+            target_id,
+            allow_deactivated=allow_deactivated,
+            allow_bots=allow_bots,
+            for_admin=for_admin,
+        )
+
+    # We're dealing with a real email address, so we need to do a careful lookup,
+    # ensuring to abide by email address visibility restrictions.
+    allowed_email_address_visibility_values = (
+        UserProfile.ROLE_TO_ACCESSIBLE_EMAIL_ADDRESS_VISIBILITY_IDS[user_profile.role]
+    )
+
     try:
-        target = get_user(email, user_profile.realm)
+        # Fetch the user from the subset of users which allow the requester to see their email address.
+        target = UserProfile.objects.get(
+            delivery_email__iexact=email.strip(),
+            realm=user_profile.realm,
+            email_address_visibility__in=allowed_email_address_visibility_values,
```

Comment:
My understanding is that we don't need a `(delivery_email, realm, email_address_visibility)` index for this query to be fast, because Postgres can figure out that it can just use the existing `(realm, delivery_email)` index to very quickly find the unique row and then simply look at  the `email_address_visibility` condition to determine if it should fetch the row or return an empty result.

Reply:
Yes, that sound correct.